File: ts/angular2/test/core/spies.ts
import {
  ChangeDetector,
  ChangeDetectorRef,
  ProtoChangeDetector,
  DynamicChangeDetector
} from 'angular2/src/core/change_detection/change_detection';

import {Renderer, RenderEventDispatcher} from 'angular2/src/core/render/api';
import {DirectiveResolver} from 'angular2/src/core/linker/directive_resolver';

import {AppView} from 'angular2/src/core/linker/view';
import {ElementRef} from 'angular2/src/core/linker/element_ref';
import {AppViewManager} from 'angular2/src/core/linker/view_manager';
import {AppViewPool} from 'angular2/src/core/linker/view_pool';
import {AppViewListener} from 'angular2/src/core/linker/view_listener';
import {ProtoViewFactory} from 'angular2/src/core/linker/proto_view_factory';
import {DomAdapter} from 'angular2/src/platform/dom/dom_adapter';
import {ClientMessageBroker} from 'angular2/src/web_workers/shared/client_message_broker';

import {
  ElementInjector,
  PreBuiltObjects,
  ProtoElementInjector
} from 'angular2/src/core/linker/element_injector';

import {SpyObject, proxy} from 'angular2/testing_internal';

export class SpyDependencyProvider extends SpyObject {}

export class SpyChangeDetector extends SpyObject {
  constructor() { super(DynamicChangeDetector); }
}

export class SpyChangeDispatcher extends SpyObject {}

export class SpyIterableDifferFactory extends SpyObject {}

export class SpyDirectiveResolver extends SpyObject {
  constructor() { super(DirectiveResolver); }
}

export class SpyView extends SpyObject {
  constructor() { super(AppView); }
}

export class SpyElementRef extends SpyObject {
  constructor() { super(ElementRef); }
}

export class SpyAppViewManager extends SpyObject {
  constructor() { super(AppViewManager); }
}

export class SpyRenderer extends SpyObject {
  constructor() {
    // Note: Renderer is an abstract class,
    // so we can't generates spy functions automatically
    // by inspecting the prototype...
    super(Renderer);
    this.spy('setEventDispatcher');
    this.spy('destroyView');
    this.spy('createView');
    this.spy('createProtoView');
    this.spy('hydrateView');
    this.spy('dehydrateView');
    this.spy('attachFragmentAfterElement');
    this.spy('attachFragmentAfterFragment');
    this.spy('detachFragment');
  }
}

export class SpyAppViewPool extends SpyObject {
  constructor() { super(AppViewPool); }
}

export class SpyAppViewListener extends SpyObject {
  constructor() { super(AppViewListener); }
}

export class SpyProtoViewFactory extends SpyObject {
  constructor() { super(ProtoViewFactory); }
}

export class SpyProtoElementInjector extends SpyObject {
  constructor() { super(ProtoElementInjector); }
}

export class SpyElementInjector extends SpyObject {
  constructor() { super(ElementInjector); }
}

export class SpyPreBuiltObjects extends SpyObject {
  constructor() { super(PreBuiltObjects); }
}

export class SpyDomAdapter extends SpyObject {
  constructor() { super(DomAdapter); }
}

export class SpyRenderEventDispatcher extends SpyObject {
  constructor() {
    // Note: RenderEventDispatcher is an interface,
    // so we can't pass it to super() and have to register
    // the spy methods on our own.
    super();
    this.spy('dispatchRenderEvent');
  }
}

File: ts/angular2/test/core/spies.dart
library core.spies;

import 'package:angular2/core.dart';
import 'package:angular2/src/core/di/injector.dart';
import 'package:angular2/src/core/change_detection/change_detection.dart';
import 'package:angular2/src/core/render/api.dart';
import 'package:angular2/src/core/linker/directive_resolver.dart';
import 'package:angular2/src/core/linker/view.dart';
import 'package:angular2/src/core/linker/element_ref.dart';
import 'package:angular2/src/core/linker/view_manager.dart';
import 'package:angular2/src/core/linker/proto_view_factory.dart';
import 'package:angular2/src/core/linker/view_pool.dart';
import 'package:angular2/src/core/linker/view_listener.dart';
import 'package:angular2/src/core/linker/element_injector.dart';
import 'package:angular2/src/platform/dom/dom_adapter.dart';
import 'package:angular2/testing_internal.dart';

@proxy
class SpyDependencyProvider extends SpyObject implements DependencyProvider {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyChangeDetector extends SpyObject implements ChangeDetector {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyChangeDispatcher extends SpyObject implements ChangeDispatcher {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyIterableDifferFactory extends SpyObject
    implements IterableDifferFactory {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyInjector extends SpyObject implements Injector {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyDirectiveResolver extends SpyObject implements DirectiveResolver {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyView extends SpyObject implements AppView {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyElementRef extends SpyObject implements ElementRef_ {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyAppViewManager extends SpyObject implements AppViewManager {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyRenderer extends SpyObject implements Renderer {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyAppViewPool extends SpyObject implements AppViewPool {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyAppViewListener extends SpyObject implements AppViewListener {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyProtoViewFactory extends SpyObject implements ProtoViewFactory {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyProtoElementInjector extends SpyObject
    implements ProtoElementInjector {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyElementInjector extends SpyObject implements ElementInjector {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyPreBuiltObjects extends SpyObject implements PreBuiltObjects {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyDomAdapter extends SpyObject implements DomAdapter {
  noSuchMethod(m) => super.noSuchMethod(m);
}

@proxy
class SpyRenderEventDispatcher extends SpyObject
    implements RenderEventDispatcher {
  noSuchMethod(m) => super.noSuchMethod(m);
}